peps/pep-0763.rst

```diff
@@ -0,0 +1,275 @@
+
+    Whenever a project detects that a particular release on PyPI might be
+    broken, they oftentimes will want to prevent further users from
+    inadvertently using that version. However, the obvious solution of
+    deleting the existing file from a repository will break users who have
+    followed best practices and pinned to a specific version of the project.
+
+    This leaves projects in a catch-22 situation where new projects may be pulling
+    down this known broken version, but if they do anything to prevent that they’ll
+    break projects that are already using it.
+
+On a technical level, the problem of deletion is mitigated by
+"yanking," also specified in :pep:`592`. However, deletions continue to be
+allowed on PyPI, and have caused multiple notable disruptions to the Python
+ecosystem over the interceding years:
+
+* July 2022: `atomicwrites <https://pypi.org/project/atomicwrites/>`_
+  was `deleted by its maintainer <https://github.com/untitaker/python-atomicwrites/issues/61>`_
+  in an attempt to remove the project's "critical" designation, without the
+  maintainer realizing that project deletion would also delete all previously
+  uploaded versions.
+
+  The project was subsequently restored with the maintainer's consent,
+  but at the cost of manual administrator action and extensive downstream
+  breakage to projects like `pytest <https://github.com/pytest-dev/pytest/issues/10114>`_.
+  As of October 2024, atomicwrites is archived but still has
+  around `4.5 million monthly downloads from PyPI <https://pypistats.org/packages/atomicwrites>`_.
+
+* April 2023: `codecov <https://pypi.org/project/codecov/>`_ was deleted by
+  its maintainers after a long deprecation period. This caused extensive
+  breakage for many of Codecov's CI/CD users, who were unaware of the
+  deprecation period due to limited observability of deprecation warnings
+  within CI/CD logs.
+
+  The project was
+  `subsequently re-created <https://about.codecov.io/blog/message-regarding-the-pypi-package/>`_
+  by its maintainers, with a new release published to compensate for the deleted releases
+  (which were not restored), meaning that any pinned installations remained
+  broken. As of October 2024, this single release remains the only release on
+  PyPI and has around
+  `1.5 million monthly downloads <https://pypistats.org/packages/codecov>`_.
+
+* June 2023: `python-sonarqube-api <https://pypi.org/project/python-sonarqube-api/>`_
+  deleted all released versions prior to 2.0.2, which was both a breaking
+  change version *and* was accompanied by a paid "Professional Edition".
+
+  The project's maintainer subsequently
+  `deleted conversations <https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227/114>`_
+  and force-pushed over the tag history for ``python-sonarqube-api``'s source repository,
+  impeding efforts by the community to compare changes between versions
+  and obtain previous, permissively-licensed versions.
```

Comment:
Is the 'paid "Professional Edition"' relevant here? Whilst this isn't great behaviour, the main point is more that older releases were deleted from the index, rather than that other methods of getting the older files were also removed. It does seem a bit harsh to single out individual cases like this in a PEP.

Reply:
I would remove the sonarqube examples. You have established a need with the other examples.

Reply:
Yeah, that's strictly not relevant. I can rephrase to emphasize the actual bits that matter (i.e. the deletion itself).